lib/Basics/FutureSharedLock.h

```diff
@@ -0,0 +1,244 @@
+      std::lock_guard lock(_mutex);
+      TRI_ASSERT(_lockCount > 0);
+      if (--_lockCount == 0 && !_queue.empty()) {
+        // we were the last lock holder -> schedule the next node
+        auto& node = _queue.back();
+        _lockCount = 1;
+        _exclusive = node->exclusive;
+        scheduleNode(node);
+        _queue.pop_back();
+
+        // if we are in shared mode, we can schedule all following shared
+        // nodes
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      }
+    }
+
+    auto insertNode(bool exclusive) {
+      _queue.emplace_front(std::make_shared<Node>(exclusive));
+      return _queue.begin();
+    }
+
+    void removeNode(
+        typename std::list<std::shared_ptr<Node>>::iterator queueIterator) {
+      TRI_ASSERT(!_queue.empty());
+      TRI_ASSERT(_lockCount > 0);
+      if (std::addressof(_queue.back()) == std::addressof(*queueIterator)) {
+        // we are the first entry in the list -> remove ourselves and check if
+        // we can schedule the next node(s)
+        _queue.pop_back();
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      } else {
+        // we are somewhere in the middle -> just remove ourselves
+        _queue.erase(queueIterator);
+      }
+    }
+
+    void scheduleExclusive(Node* node) { scheduleNode(node); }
+
+    void scheduleShared(Node* leader) { scheduleNode(leader); }
+
+    void scheduleNode(std::shared_ptr<Node> node) {
+      TRI_ASSERT(node != nullptr);
+      node->scheduled = true;
+      _scheduler.queue(
+          [node, this]() { node->promise.setValue(LockGuard(this)); });
+    }
+
+    void scheduleTimeout(
+        typename std::list<std::shared_ptr<Node>>::iterator queueIterator,
+        std::chrono::milliseconds timeout) {
+      _scheduler.queueDelayed(
+          [self = this->weak_from_this(),
+           node = std::weak_ptr<Node>(*queueIterator),
+           queueIterator]() mutable {
+            auto me = self.lock();
+            if (me) {
+              std::lock_guard lock(me->_mutex);
+              auto nodePtr = node.lock();
+              if (nodePtr != nullptr && !nodePtr->scheduled) {
+                // if we could lock the node, the iterator must still be valid!
+                me->removeNode(queueIterator);
+                TRI_ASSERT(nodePtr.get() != nullptr) << (void*)nodePtr.get();
+              }
+            }
+          },
+          timeout);
+    }
+
+    Scheduler& _scheduler;
+    std::mutex _mutex;
+    // the list is ordered back to front, i.e., new items are added at the
+    // front; the last item (back) is the next to be scheduled
+    std::list<std::shared_ptr<Node>> _queue;
```

Comment:
It can be intrusive list

Reply:
Yes, it could be an intrusive list, but for now it was simpler this way.

Reply:
Ok